### PR TITLE
Allow paramters to be ignored in Stubber validation

### LINKED ISSUE #931: Feature request:  Add ability to ignore params in stubber
It would be nice if the stubber had a solution for being able to ignore randomly generated values in the input.  For example:

* idempotency tokens
* statement ids

And really anything that typically is a random value.  

There's a few ways I can think of to accomplish this.  In `mock`, you can use `mock.ANY` to indicate that you don't care about the value.   In the CLI we have an `ignore_params` for `assert_params_for_cmd`, though this only works for top level params (which typically is sufficient).

### PR DESCRIPTION
**Pull request by JordonPhillips**
This allows users to specify stub.ANY for expected parameters if
they don't care about a specific parameter's value.

Resolves #931

cc @kyleknap  @jamesls

### CODE REVIEW

File: botocore/stub.py [hunk 1]
```diff
@@ -265,7 +314,21 @@ def _get_response_handler(self, model, params, **kwargs):
...
         expected_params = self._queue[0]['expected_params']
-        if expected_params is not None and params != expected_params:
+        if expected_params is None:
+            return
+
+        # Validate the parameters are equal
+        for param, value in expected_params.items():
+            if value is ANY:
+                continue
+            elif expected_params[param] != params[param]:
```
Comment (kyleknap) [inline botocore/stub.py]: I feel like there may be key errors here if the expected param is missing from the actual params.

File: botocore/stub.py [hunk 2]
```diff
@@ -265,7 +314,21 @@ def _get_response_handler(self, model, params, **kwargs):
...
+                raise StubAssertionError(
+                    operation_name=model.name,
+                    reason='Expected parameters:\n%s,\nbut received:\n%s' % (
```
Comment (kyleknap) [inline botocore/stub.py]: I wonder if we can make the error message much cleaner in telling what parameters are differing and how instead of having to compare the dictionaries because I actually find that a bit annoying.

**Comment by kyleknap** [comment]
I like the idea. Allows me to remove mock completely from ``s3transfer``. But had a couple comments.

File: botocore/stub.py [hunk 3]
```diff
@@ -20,6 +20,9 @@
 from botocore.vendored.requests.models import Response
 
 
+ANY = object
```
Comment (jamesls) [inline botocore/stub.py]: `s/object/object()`.

File: botocore/stub.py (hunk 2 above)
Comment (jamesls) [inline botocore/stub.py]: I agree but I think that should be a separate PR right?

File: tests/functional/test_stub.py [hunk 4]
```diff
@@ -148,3 +150,19 @@ def test_still_relies_on_param_validation_with_expected_params(self):
         # Throw an error for invalid parameters
         with self.assertRaises(ParamValidationError):
             self.client.list_objects(Buck='bar')
+
+    def test_any_ignores_param_for_validation(self):
```
Comment (jamesls) [inline tests/functional/test_stub.py]: Can we add a test with multiple params?  One stub.ANY and one with an expected value and verify you're allowed to mix the two types in the same response call?

**Comment by jamesls** [comment]
Just had some small feedback.

File: botocore/stub.py (hunk 2 above)
Comment (kyleknap) [inline botocore/stub.py]: Yeah probably more appropriate in separate PR.

**Comment by JordonPhillips** [comment]
Updated

File: botocore/stub.py [hunk 5]
```diff
@@ -265,7 +314,21 @@ def _get_response_handler(self, model, params, **kwargs):
...
         expected_params = self._queue[0]['expected_params']
-        if expected_params is not None and params != expected_params:
+        if expected_params is None:
+            return
+
+        # Validate the parameters are equal
+        for param, value in expected_params.items():
+            if value is ANY:
+                continue
+            elif expected_params[param] != params.get(param, None):
```
Comment (jamesls) [inline botocore/stub.py]: I think it's useful to distinguish between explicitly set values versus missing values.  I'd expect something like this to trigger an assertion error:

`{"foo": None} == {}`

Comment (kyleknap) [inline botocore/stub.py]: I had the same comment.

**Comment by JordonPhillips** [comment]
updated again

**Comment by kyleknap** [comment]
Thanks. :ship:

**Comment by jamesls** [comment]
Where's the test that covers the last condition @kyleknap and myself asked you to address?  I'm not seeing it.

If you haven't added one, then as a reminder, you should be adding tests for all the various cases you're handling.

**Comment by JordonPhillips** [comment]
Seemed at the time to be testing an implementation detail, but I realize now it really isn't. Updated.

**Comment by jamesls** [comment]
:shipit:
